#  Modificando Colunas e Valores

### Vamos pegar o nosso dataframe novamente

In [147]:
import pandas as pd
#importando os arquivos
vendas_df = pd.read_csv(r'Contoso - Vendas  - 2017.csv', sep=';', encoding="ISO-8859-1" )
produtos_df = pd.read_csv(r'Contoso - Cadastro Produtos.csv', sep=';', encoding="ISO-8859-1")
lojas_df = pd.read_csv(r'Contoso - Lojas.csv', sep=';', encoding="ISO-8859-1")
clientes_df = pd.read_csv(r'Contoso - Clientes.csv', sep=';' , encoding="ISO-8859-1")

clientes_df = clientes_df.rename(columns={"ÿID Cliente": "ID Cliente"})
clientes_df = clientes_df.rename(columns={"E-mail": "E-mail Cliente"})
lojas_df = lojas_df.rename(columns={"ÿID Loja": "ID Loja"})
produtos_df = produtos_df.rename(columns={"ÿNome do Produto": "Nome do Produto"})

#limpando apenas as colunas que queremos
clientes_df = clientes_df[['ID Cliente', 'E-mail Cliente']]
produtos_df = produtos_df[['ID Produto', 'Nome do Produto']]
lojas_df = lojas_df[['ID Loja', 'Nome da Loja']]

#mesclando e renomeando os dataframes
vendas_df = vendas_df.merge(produtos_df, on='ID Produto')
vendas_df = vendas_df.merge(lojas_df, on='ID Loja')
vendas_df = vendas_df.merge(clientes_df, on='ID Cliente')
display(vendas_df)

Numero da Venda Data da Venda Data do Envio  ID Canal  ID Loja  \
0                     1    01/01/2017    02/01/2017         1       86   
1                     2    01/01/2017    06/01/2017         5      308   
2                     3    01/01/2017    01/01/2017         0      294   
3                     4    01/01/2017    01/01/2017         0      251   
4                     5    01/01/2017    07/01/2017         6       94   
...                 ...           ...           ...       ...      ...   
980637           980638    31/12/2017    31/12/2017         0      194   
980638           980639    31/12/2017    06/01/2018         6       32   
980639           980640    31/12/2017    03/01/2018         3      210   
980640           980641    31/12/2017    31/12/2017         0       53   
980641           980642    31/12/2017    31/12/2017         0      309   

        ID Produto  ID Promocao  ID Cliente  Quantidade Vendida  \
0              981            2        6825                   9   
1             1586            2       18469                   9   
2             1444            5       19730                  13   
3             1468            5       29326                   6   
4             1106            2       22617                   4   
...            ...          ...         ...                 ...   
980637        2490            4       10353                 120   
980638        2488            4       31750                 120   
980639        2511           10       12003                  80   
980640         436            4       25550                  39   
980641        2510           10       28707                  80   

        Quantidade Devolvida  \
0                          1   
1                          1   
2                          1   
3                          1   
4                          1   
...                      ...   
980637                     0   
980638                     0   
980639                     0   
980640                     0   
980641                     0   

                                          Nome do Produto  \
0              A. Datum Advanced Digital Camera M300 Pink   
1                   SV DVD 55DVD Storage Binder M56 Black   
2       The Phone Company Touch Screen Phones 26-2.2" ...   
3             Contoso Touch Screen Phones - CRT M11 Black   
4                          Contoso SLR Camera M146 Orange   
...                                                   ...   
980637      Contoso Rubberized Skin BlackBerry E100 White   
980638      Contoso Rubberized Skin BlackBerry E100 Black   
980639  Contoso Original K1m Li-Ion Standard Battery E...   
980640         Adventure Works Desktop PC1.80 ED180 White   
980641  Contoso Original K1m Li-Ion Standard Battery E...   

                               Nome da Loja                E-mail Cliente  
0                      Loja Contoso Austin         rbrumfieldmy@ameblo.jp  
1       Loja Contoso North America Reseller        cshawd4@technorati.com  
2                 Loja Contoso Tehran No.2         kgorriekd@bandcamp.com  
3                   Loja Contoso Singapore   angela49@adventure-works.com  
4               Loja Contoso Grand Prairie     jacob4@adventure-works.com  
...                                     ...                           ...  
980637                Loja Contoso Ridgely        ocrosson9u@illinois.edu  
980638               Loja Contoso Milliken   victor11@adventure-works.com  
980639               Loja Contoso Cheshire               mtopaz4@gmpg.org  
980640              Loja Contoso East Troy     lance5@adventure-works.com  
980641         Loja Contoso Europe Reseller      jan8@adventure-works.com  

[980642 rows x 13 columns]

In [148]:
vendas_df["Data da Venda"] = pd.to_datetime(vendas_df["Data da Venda"], format="%d/%m/%Y")
vendas_df["Data do Envio"] = pd.to_datetime(vendas_df["Data do Envio"], format="%d/%m/%Y")

vendas_df["Ano da Venda"] = vendas_df["Data da Venda"].dt.year
vendas_df["Mês da Venda"] = vendas_df["Data da Venda"].dt.month
vendas_df["dia da Venda"] = vendas_df["Data da Venda"].dt.day
vendas_df["Ano do Envio"] = vendas_df["Data do Envio"].dt.year
vendas_df["Mês do Envio"] = vendas_df["Data do Envio"].dt.month
vendas_df["Dia do Envio"] = vendas_df["Data do Envio"].dt.day
display(vendas_df)

Numero da Venda Data da Venda Data do Envio  ID Canal  ID Loja  \
0                     1    2017-01-01    2017-01-02         1       86   
1                     2    2017-01-01    2017-01-06         5      308   
2                     3    2017-01-01    2017-01-01         0      294   
3                     4    2017-01-01    2017-01-01         0      251   
4                     5    2017-01-01    2017-01-07         6       94   
...                 ...           ...           ...       ...      ...   
980637           980638    2017-12-31    2017-12-31         0      194   
980638           980639    2017-12-31    2018-01-06         6       32   
980639           980640    2017-12-31    2018-01-03         3      210   
980640           980641    2017-12-31    2017-12-31         0       53   
980641           980642    2017-12-31    2017-12-31         0      309   

        ID Produto  ID Promocao  ID Cliente  Quantidade Vendida  \
0              981            2        6825                   9   
1             1586            2       18469                   9   
2             1444            5       19730                  13   
3             1468            5       29326                   6   
4             1106            2       22617                   4   
...            ...          ...         ...                 ...   
980637        2490            4       10353                 120   
980638        2488            4       31750                 120   
980639        2511           10       12003                  80   
980640         436            4       25550                  39   
980641        2510           10       28707                  80   

        Quantidade Devolvida  \
0                          1   
1                          1   
2                          1   
3                          1   
4                          1   
...                      ...   
980637                     0   
980638                     0   
980639                     0   
980640                     0   
980641                     0   

                                          Nome do Produto  \
0              A. Datum Advanced Digital Camera M300 Pink   
1                   SV DVD 55DVD Storage Binder M56 Black   
2       The Phone Company Touch Screen Phones 26-2.2" ...   
3             Contoso Touch Screen Phones - CRT M11 Black   
4                          Contoso SLR Camera M146 Orange   
...                                                   ...   
980637      Contoso Rubberized Skin BlackBerry E100 White   
980638      Contoso Rubberized Skin BlackBerry E100 Black   
980639  Contoso Original K1m Li-Ion Standard Battery E...   
980640         Adventure Works Desktop PC1.80 ED180 White   
980641  Contoso Original K1m Li-Ion Standard Battery E...   

                               Nome da Loja                E-mail Cliente  \
0                      Loja Contoso Austin         rbrumfieldmy@ameblo.jp   
1       Loja Contoso North America Reseller        cshawd4@technorati.com   
2                 Loja Contoso Tehran No.2         kgorriekd@bandcamp.com   
3                   Loja Contoso Singapore   angela49@adventure-works.com   
4               Loja Contoso Grand Prairie     jacob4@adventure-works.com   
...                                     ...                           ...   
980637                Loja Contoso Ridgely        ocrosson9u@illinois.edu   
980638               Loja Contoso Milliken   victor11@adventure-works.com   
980639               Loja Contoso Cheshire               mtopaz4@gmpg.org   
980640              Loja Contoso East Troy     lance5@adventure-works.com   
980641         Loja Contoso Europe Reseller      jan8@adventure-works.com   

        Ano da Venda  Mês da Venda  dia da Venda  Ano do Envio  Mês do Envio  \
0               2017             1             1          2017             1   
1               2017             1             1          2017             1   
2               2017           

### Agora, e se quisermos acrescentar uma coluna com o mês, o dia e o ano de cada venda (e não só a data completa)

### E agora, caso a gente queira modificar 1 valor específico, como fazemos? Vamos importar novamente a base de produtos

In [149]:


novo_produtos_df = pd.read_csv(r'Contoso - Cadastro Produtos.csv', sep=';', encoding="ISO-8859-1")
novo_produtos_df = novo_produtos_df.rename(columns={"ÿNome do Produto": "Nome do Produto"})
novo_produtos_df.iloc[3,6] = 50
display(novo_produtos_df.head())
#repare no .head() para pegar apenas os primeiros valores, é bem comum esse uso para ter uma visão do que são os dados

Nome do Produto  \
0            Contoso Wireless Laser Mouse E50 Grey   
1    Contoso Optical Wheel OEM PS/2 Mouse E60 Grey   
2   Contoso Optical Wheel OEM PS/2 Mouse E60 Black   
3   Contoso Optical Wheel OEM PS/2 Mouse E60 White   
4  Contoso Optical Wheel OEM PS/2 Mouse E60 Silver   

                                Descricao do Produto    Fabricante  \
0  Advanced 2.4 GHz cordless technology makes fre...  Contoso, Ltd   
1                     PS/2 mouse, 6 feet mouse cable  Contoso, Ltd   
2                     PS/2 mouse, 6 feet mouse cable  Contoso, Ltd   
3                     PS/2 mouse, 6 feet mouse cable  Contoso, Ltd   
4                     PS/2 mouse, 6 feet mouse cable  Contoso, Ltd   

  Nome da Marca       Tipo Custo Unitario Preco Unitario  ID Produto  \
0       Contoso  Econômico          10,69          20,96         873   
1       Contoso  Econômico           6,63             13         879   
2       Contoso  Econômico           6,63             13         880   
3       Contoso  Econômico           6,63             50         881   
4       Contoso  Econômico           6,63             13         882   

   ID Subcategoria  
0               22  
1               22  
2               22  
3               22  
4               22

### Antes de entrar no próximo exemplo, precisamos falar de 2 métodos:
    1. loc - permite pegar uma linha de acordo com o índice dela. Ele dá erro caso não encontre o índice. Isso é interessante principalmente quando o índice é uma informação relevante ao invés só do número do índice ou quando queremos pegar alguma linha específica do dataframe (ao invés de ir do início do dataframe até a linha 5, por exemplo).
        Também podemos usar como loc[índice_linha, índice_coluna] para acessar um valor específico e modificá-lo.
    2. iloc - enxerga o dataframe como linhas e colunas e consegue pegar o valor com um número de linha e um número de coluna. Repara que ele não analisa o valor do índice da linha e da coluna, apenas a posição importa.
        Uso: iloc[num_linha, num_coluna]
        
- Vendo na prática

In [150]:
novo_produtos_df = novo_produtos_df.set_index('Nome do Produto')
display(novo_produtos_df.head())

Descricao do Produto  \
Nome do Produto                                                                                      
Contoso Wireless Laser Mouse E50 Grey            Advanced 2.4 GHz cordless technology makes fre...   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey                       PS/2 mouse, 6 feet mouse cable   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black                      PS/2 mouse, 6 feet mouse cable   
Contoso Optical Wheel OEM PS/2 Mouse E60 White                      PS/2 mouse, 6 feet mouse cable   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver                     PS/2 mouse, 6 feet mouse cable   

                                                   Fabricante Nome da Marca  \
Nome do Produto                                                               
Contoso Wireless Laser Mouse E50 Grey            Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey    Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black   Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 White   Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver  Contoso, Ltd       Contoso   

                                                      Tipo Custo Unitario  \
Nome do Produto                                                             
Contoso Wireless Laser Mouse E50 Grey            Econômico          10,69   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey    Econômico           6,63   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black   Econômico           6,63   
Contoso Optical Wheel OEM PS/2 Mouse E60 White   Econômico           6,63   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver  Econômico           6,63   

                                                Preco Unitario  ID Produto  \
Nome do Produto                                                              
Contoso Wireless Laser Mouse E50 Grey                    20,96         873   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey               13         879   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black              13         880   
Contoso Optical Wheel OEM PS/2 Mouse E60 White              50         881   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver             13         882   

                                                 ID Subcategoria  
Nome do Produto                                                   
Contoso Wireless Laser Mouse E50 Grey                         22  
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey                 22  
Contoso Optical Wheel OEM PS/2 Mouse E60 Black                22  
Contoso Optical Wheel OEM PS/2 Mouse E60 White                22  
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver               22

In [151]:
#vamos pegar o preço produto Contoso Optical Wheel OEM PS/2 Mouse E60 Black
#por loc
#novo_produtos_df = novo_produtos_df.loc["Contoso Optical Wheel OEM PS/2 Mouse E60 Black", "Preco Unitario"]
print(novo_produtos_df.loc["Contoso Optical Wheel OEM PS/2 Mouse E60 Black", "Preco Unitario"])
#por iloc
print(novo_produtos_df.iloc[2, 5])



13
13


### A empresa decidiu aumentar o preço do produto ID 873 (Contoso Wireless Laser Mouse E50 Grey) para 23 reais. Como fazemos, para modificar isso na nossa base?

In [152]:
novo_produtos_df.loc['Contoso Wireless Laser Mouse E50 Grey', 'Preco Unitario'] = 23
novo_produtos_df.loc['Contoso Optical Wheel OEM PS/2 Mouse E60 Silver', 'Preco Unitario'] = 45

    

display(novo_produtos_df.head())

Descricao do Produto  \
Nome do Produto                                                                                      
Contoso Wireless Laser Mouse E50 Grey            Advanced 2.4 GHz cordless technology makes fre...   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey                       PS/2 mouse, 6 feet mouse cable   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black                      PS/2 mouse, 6 feet mouse cable   
Contoso Optical Wheel OEM PS/2 Mouse E60 White                      PS/2 mouse, 6 feet mouse cable   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver                     PS/2 mouse, 6 feet mouse cable   

                                                   Fabricante Nome da Marca  \
Nome do Produto                                                               
Contoso Wireless Laser Mouse E50 Grey            Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey    Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black   Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 White   Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver  Contoso, Ltd       Contoso   

                                                      Tipo Custo Unitario  \
Nome do Produto                                                             
Contoso Wireless Laser Mouse E50 Grey            Econômico          10,69   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey    Econômico           6,63   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black   Econômico           6,63   
Contoso Optical Wheel OEM PS/2 Mouse E60 White   Econômico           6,63   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver  Econômico           6,63   

                                                Preco Unitario  ID Produto  \
Nome do Produto                                                              
Contoso Wireless Laser Mouse E50 Grey                       23         873   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey               13         879   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black              13         880   
Contoso Optical Wheel OEM PS/2 Mouse E60 White              50         881   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver             45         882   

                                                 ID Subcategoria  
Nome do Produto                                                   
Contoso Wireless Laser Mouse E50 Grey                         22  
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey                 22  
Contoso Optical Wheel OEM PS/2 Mouse E60 Black                22  
Contoso Optical Wheel OEM PS/2 Mouse E60 White                22  
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver               22

In [153]:
# quando queremos aplicar o valor unico para todos os elementos ou produtos
# for i, produto in enumerate(novo_produtos_df):
   
#     novo_produtos_df.iloc[i,5] = 50

# display(novo_produtos_df.head())

In [154]:
# Aplicar o preço em porcentagem para todos os produtos (adicionando mais valor sobre o antigo)

# Converte a coluna 'Preco Unitario' para o tipo numérico
novo_produtos_df['Preco Unitario'] = pd.to_numeric(novo_produtos_df['Preco Unitario'], errors='coerce')

# Verifica se a conversão foi bem-sucedida
novo_produtos_df['Preco Unitario'] = novo_produtos_df['Preco Unitario'].fillna(0)  # Define NaNs como 0, caso haja valores não numéricos

# Adiciona 10 ao preço unitário de cada produto
for i in range(len(novo_produtos_df)):
    novo_produtos_df.iloc[i, 5] *= 1.1  # Agora deve funcionar corretamente

# Exibindo o DataFrame atualizado
display(novo_produtos_df.head())


Descricao do Produto  \
Nome do Produto                                                                                      
Contoso Wireless Laser Mouse E50 Grey            Advanced 2.4 GHz cordless technology makes fre...   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey                       PS/2 mouse, 6 feet mouse cable   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black                      PS/2 mouse, 6 feet mouse cable   
Contoso Optical Wheel OEM PS/2 Mouse E60 White                      PS/2 mouse, 6 feet mouse cable   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver                     PS/2 mouse, 6 feet mouse cable   

                                                   Fabricante Nome da Marca  \
Nome do Produto                                                               
Contoso Wireless Laser Mouse E50 Grey            Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey    Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black   Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 White   Contoso, Ltd       Contoso   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver  Contoso, Ltd       Contoso   

                                                      Tipo Custo Unitario  \
Nome do Produto                                                             
Contoso Wireless Laser Mouse E50 Grey            Econômico          10,69   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey    Econômico           6,63   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black   Econômico           6,63   
Contoso Optical Wheel OEM PS/2 Mouse E60 White   Econômico           6,63   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver  Econômico           6,63   

                                                 Preco Unitario  ID Produto  \
Nome do Produto                                                               
Contoso Wireless Laser Mouse E50 Grey                      25.3         873   
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey              14.3         879   
Contoso Optical Wheel OEM PS/2 Mouse E60 Black             14.3         880   
Contoso Optical Wheel OEM PS/2 Mouse E60 White             55.0         881   
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver            49.5         882   

                                                 ID Subcategoria  
Nome do Produto                                                   
Contoso Wireless Laser Mouse E50 Grey                         22  
Contoso Optical Wheel OEM PS/2 Mouse E60 Grey                 22  
Contoso Optical Wheel OEM PS/2 Mouse E60 Black                22  
Contoso Optical Wheel OEM PS/2 Mouse E60 White                22  
Contoso Optical Wheel OEM PS/2 Mouse E60 Silver               22

In [155]:
novo_produtos_df.to_csv("Meu_Arquivo_final.csv", sep=";")